### ISSUE TITLE
Reorganizing the code tree.

### ISSUE THREAD

**Issue by Arthur-Milchior** [ankidroid/Anki-Android#12582]
I'd love to do some cleanup in the code tree. I find it very confusing, even after almost 3 years working in it.

Right now, this issue is up for discussion, we are not expecting anyone to do anything until there is a consensus. The first question is whether you agree that it's, on principle, something worth doing, whether you oppose it, or whether you don't care (I'm not asking whether you want to work on it yourself, I'm fine doing it myself, or getting help of whoever people also belive it'll improve the state of the codebase and contributors work)

If we agree that it's worth doing a change, I suggest a new organization of the tree in https://docs.google.com/document/d/1IfEdbSWqWdpWWZUkFiENqty7twAupyJou7JakDvHkEo/edit?usp=sharing
I believe a design document is better than an issue because anyone can comment any specific point, and we can track discussions of each directory proposal independently.

**Comment by n1snt** [comment]
I strongly agree with the 2nd point you have mentioned in the doc to add a README.md in each directory.
This could help a lot of new contributors know what is located where.

**Comment by divyansh-dxn** [comment]
One more common approach is to have packages divided into data-domain-ui layers. 
The data package would have the codes related to sqlite and backend
The mid processing like mapping, computing something like deck tree, building cache, etc can be put in domain 
And UI as name suggests would be having all the activities and fragments. Maybe at this point of time we can include view models as well

Sorry I could document it properly...but yeah just wanted to give some idea. For my personal projects I follow the same principle.

**Comment by Arthur-Milchior** [comment]
> Sorry I could document it properly.

I don't understand what you are sorry about. I'd be extremely happy to read other suggestions; I want this to occur only once by decade; so let's do it right. This can even wait for after GSoC, as I know you have other priorities right now.

Clearly, right now, I'm biaised by work, where each folder contains together the controler, the coordinator, the delegate(s) if any, and the mediator. But it's only one design choice that may not be ideal everywhere.

I must admit I fear your suggestion requires to edit a lot of files, which are not well separated between UI and processing; but that's maybe something that we should do to improve things. Just, I 'd love to see how we proceed to decide whether `settings` for example, are only `ui` or whether it's processing ; because it does some itself.
Also, where would you put all of AsyncTask and the coroutine architecture; all of the exceptions. Is `Time` part of data or domain; I find Time to be very low level, and we manipulate it for testing; but it's clearly not part of our domain.

Reciprocally, when you have time, I guess I'd love to see the intended content of those three folders. As you probably won't put everything at top level in one of those folders, that seems interesting to project myself in this project if I can visualize at least one example of domain subfolder, one example of ui subfolder

**Comment by thedroiddiv** [comment]
keep open

**Comment by n1snt** [comment]
Can anyone please add the `Keep Open` label to this issue?

**Comment by oakkitten** [comment]
I wonder if it would be more useful to separate stuff based primarily on the role it performs in the application, rather than its type or domain. For example, “storage migration” encompasses various elements such as the migration executor, the service, and UI dialogs. I think that perhaps it would be best to put them all, or as much as you can—which can usually be a lot if you put your mind to it—into `com.ichi2.anki.migration`. Why? Well, with the way AS works I never consider importing an issue. Sure, you might have have to e.g. import `com.ichi2.anki.dialogs.tools` (generic dialog tools) in a dialog that resides in `com.ichi2.anki.migration`—but you are just pressing Alt-Enter or whatever, no big deal. But:
  * when working on scoped storage, you will have everything together—easy navigation, and
  * as a developer, you get incentivized to put related stuff together, and what's more importantly, to put it outside the megaclasses such as `DeckPicker`. Got a _thing_ that's related to some role but that doesn't fit into any existing types? Well, at least you know where to put it!

**Comment by dae** [comment]
I'd suggest the first step be to split libanki off into a separate library (in the same repo, but a separate top-level folder), which the GUI code imports. There are probably incremental compilation and testing performances wins to be had by doing so, and libanki should no longer require Robolectric: https://github.com/ankidroid/Anki-Android/pull/14171/commits/0af96c791eab2749106a8ca0ccc6561ce43ff129

**Comment by lukstbit** [comment]
My opinion would be to split code by functionality:

`com.ichi2.anki` : the root package for all modules(with the exception of `AnkiDroid` each package will of course add its own suffix like api, rules etc)


`Structure:`
- `module:AnkiDroid` with the following package structure(basic, no particular order):
   - `analytics` : self explanatory
   - `startup` : contains onboarding and other starting related activities/code(ex. IntentHandler)
   - `cardbrowser` : remains native, will contain all code(fragments,dialogs etc) related
   - `deckpicker` : remains native, will contain all code(fragments,dialogs etc) related 
   - `settings/preferences` : will contain all current code related to preferences(we currently have 2 now for these)
   - `jsaddons` : not sure about this, keep it as it is?!
   - `notetypes` : remains native?!, current code around managing the noteytypes and their fields
   - `provider` : the contentProvider remains as it is now
   - `notifications` : everything related to notifications
   - `widget` : very small amount of code, but we could consider moving it to its own module(this will allow us to test modern libraries      like compose glance in isolation from main code)
   - `storage` : contains the migration package(easier to delete it when minimum is Android 10), contains the import/export features as well
   - `core` : code around the current _pages_ package, the screens that Brayan wants to migrate will go here(own packages?)
  - `utilities/support` : contains all current utilities(need cleanup(a lot of dead code) and to consolidate them) + current _compat_ package. These would be rarely changed and we could move it to their own module.
  - `common` : this would contain code that is common in packages above(an example would be the deck selection dialog which is common in CardBrowser and DeckPicker). The code in this package could be put in `utilities/support` but I would rather have that package as a pure utilities package and not mix it with app logic.

 
- `module:libanki` : current code in the package libanki(Note that I've looked at this(not the related tests) and cleaned up that package for cross references and android related classes with the exception of Collection.kt and Sound.kt and should be easy to move around)
- `module:api` : only changing the package structure inside to have everything under _com.ichi2.anki.api_
- `module:lint-rules` : remains as it is now
- `module:build-logic` : some plugins that I plan to write and other build related stuff to cleanup the build system

**Comment by vaibhavsawant5** [comment]
I have been looking at the code only since a week and find it very confusing to understand. I have trying to figure out how to refactor the code(not the package structure) and I have some suggestions to start with
1. Start with DeckPicker and keep it as parent activity within which all other fragments open.
2. Extract all code from DeckPicker into a fragment e.g. DeckPickerFragment.
3. Make CardBrowser, Statistics and Preferences activities as fragments so they don't open in new screens. Instead, populate the existing space.

Starting with these can help a newbie like me to understand the starting code better.

**Comment by david-allison** [comment]
As mentioned in https://github.com/ankidroid/Anki-Android/pull/16498, splitting out `compat` feels like a next logical step (besides the CustomTabs logic).

This also a good opportunity to rename the package:

* com.ichi2.compat -> com.ichi2.anki.compat
